### PR TITLE
Create ActionStats for completed voter registrations

### PR DESCRIPTION
**Pull request by aaronschachter**
### What's this PR do?

This pull request creates or updates an `ActionStat` for a school upon creating or updating a completed `voter-reg` post. We'll eventually use this data to display a leaderboard of voter registrations by school (details in #[2417735](https://www.pivotaltracker.com/n/projects/2417735)).

It implements this by refactoring the existing Post `getAcceptedQuantitySum` method as `updateOrCreateActionStats`, and by calling it from newly added `created` and `updated` hooks in the PostObserver.

Also adds a static `getCompletedVoterRegStatuses` method to the Post model to DRY getting list of statuses considered completed.

### How should this be reviewed?

👀 

### Any background context you want to provide?

* We have existing test coverage for creating ActionStats for approved posts in test for the Reviews endpoint, so didn't add it into our model test for now. Might be worth moving it later to keep the tests in one place.

* A followup PR will add a `state` field to the `action_stats` table, in order to filter action stats by state. Holding on adding that in this PR to hopefully keep this easier to review.

### Relevant tickets

References [Pivotal #2417735](https://www.pivotaltracker.com/n/projects/2417735).

### Checklist

- [x] This PR has been added to the relevant Pivotal card.
- [ ] Documentation added for new features/changed endpoints.
- [ ] Added appropriate feature/unit tests.

### CODE REVIEW

File: app/Repositories/PostRepository.php [hunk 1]
```diff
@@ -205,14 +204,6 @@ public function reviews(Post $post, $status, $comment = null, $admin = null)
         // Update the "counter cache" on the Post Campaign:
         $post->campaign->refreshCounts();
 
-        // If the Post has a School ID, upsert an ActionStat.
-        if ($post->school_id && $post->action_id) {
```
Comment (mendelB) [inline app/Repositories/PostRepository.php:209]: Why were we checking for the `action_id` here? I noticed we omit that check above?

**Comment by mendelB** [comment]
Random Q -- did we ever intend to start queuing the `ActionStat` generation, or is their performance trivial?

> A followup PR will add a state field to the action_stats table, in order to filter action stats by state. Holding on adding that in this PR to hopefully keep this easier to review.

Also lol does this mean like 'status' `state` or the post location? (assuming the former, but that did confuse me since we have both on posts 😅 )

**Comment by aaronschachter** [comment]
Re: `state`... I do mean post location. I did think about calling `school_state`... but I have a feeling we eventually may want to refactor the `action_stats` table to support storing action stats for groups -- right now we only store by schooll. I think it'd be a mess to have both `school_state` and `group_state` fields, since they'd both be identical... but agree that `state` isn't exactly crystal clear ...

File: app/Repositories/PostRepository.php (hunk 1 above)
Comment (aaronschachter) [reply app/Repositories/PostRepository.php:209]: I think this was maybe to safeguard for when posts are created with an `action` vs `action_id` -- but I'm not sure this was even necessary.

File: app/Models/Post.php [hunk 2]
```diff
@@ -590,21 +590,51 @@ public static function getPostCount(Campaign $campaign, string $status)
...
+        // We currently only save action stats for posts that belong to schools.
+        if (! $this->school_id) {
+            return;
+        }
+
+        // If completed voter-reg post, update school impact as completed count for this action.
+        if ($this->type === 'voter-reg' && in_array($this->status, self::getCompletedVoterRegStatuses())) {
+            $impact = (new self)->newModelQuery()
+                ->where('school_id', $this->school_id)
+                ->where('type', 'voter-reg')
```
[the remark this answers is not in the document]
Comment (aaronschachter) [reply app/Models/Post.php:608]: I think in theory yes -- but the action ID might change for voter reg posts per election year.

**Comment by aaronschachter** [comment]
> Random Q -- did we ever intend to start queuing the ActionStat generation, or is their performance trivial?

I think for now, it's not too often we see posts with school ID's, so I haven't been too concerned... cc @DFurnes

**Comment by mendelB** [comment]
> Re: `state`... I do mean post location. I did think about calling `school_state`... but I have a feeling we eventually may want to refactor the `action_stats` table to support storing action stats for groups -- right now we only store by schooll. I think it'd be a mess to have both `school_state` and `group_state` fields, since they'd both be identical... but agree that `state` isn't exactly crystal clear ...

Ahh fair. Hmm I suppose `location_state` isn't exactly elegant either 🤔  We call this simply `location` on Posts, which we do have filters for in Phoenix galleries, do you think staying consistent is a win?

**Comment by aaronschachter** [comment]
I see that the posts `location` field contains values like `US-VT` -- but we store `VT` in a `state` field on both the Group Rogue [resource](https://github.com/DoSomething/rogue/blob/master/docs/endpoints/groups.md) and [schools](https://github.com/DoSomething/graphql/blob/9e17f767e24414f89c4382038a92833cec43d7ae/src/schema/schools.js#L21), we also allow filtering by it too. I kind of lean towards `state` being more consistent vs `location`, because of the data and existing `state` filters?

**Comment by DFurnes** [comment]
For naming, I think I'd recommend using `location` for consistency (and because, like Mendel mentioned, `state` has a lot of different meanings). That said, I totally get that `location` is maybe a little to vague! If I were renaming the `location` field today, I think I'd call it `region` since these are ISO-3166-2 region codes (and not necessarily _just_ US states).

**Comment by DFurnes** [comment]
I don't know if it's too late, but I'd love to store the group/school locations as ISO-3166-2 codes as well for consistency.

**Comment by DFurnes** [comment]
> I think for now, it's not too often we see posts with school ID's, so I haven't been too concerned... cc @DFurnes

Yeah, I think keeping things simple & then optimizing when we see a performance issue seems smart! I think right now the biggest performance outlier in this application (ugh) is re-calculating post counts on the `/api/v3/reviews` endpoint.

**Comment by aaronschachter** [comment]
> I don't know if it's too late, but I'd love to store the group/school locations as ISO-3166-2 codes as well for consistency.

It's too late for Monday's launch, both the Great Schools Mongo DB and our `state` field in the Rogue `groups` table are storing the US codes. It'd require corresponding changes in GraphQL and the `UsaStateSelect` Phoenix component too. A good cleanup task would be to update all data in each field and rename.

**Comment by mendelB** [comment]
So are you thinking keeping this as `state` pending an epic cleanup across the batch of 'em?

**Comment by aaronschachter** [comment]
I'm thinking of going with `location`, and storing it as ISO-3166-2. It's one less to cleanup later. Changing the `state` field name in our `Group` model isn't terribly tricky since we can create a new field called `location` and migrate state data over as   ISO-3166-2, so maybe I'll sneak it into the next PR. Eventually we can switch the GraphQL queries to query by `location` instead of `state` (for GreatSchools, we could get away with formatting the field in GraphQL vs running a db query.. which would keep the DB update process simple)

**Comment by mendelB** [comment]
Ahh wicked neat, that sounds like a smart way to go!!

**Pull request merged**